File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    /*int lengthOfLongestSubstring(string s) {
       int ans=0,l=0,r=0;//sliding window approacch
        vector<int>store(256,0);
        while(r<s.length()){
            store[s[r]]++;
            while(store[s[r]]>1){
                store[s[l]]--;//erases the occurences of letters in the current window 
                l++;
            }
            ans=max(ans,r-l+1);
            r++;
        }
    return ans;}*/
    int lengthOfLongestSubstring(string s){
        int n=s.length(),i=0,j=0,ans=0;
        if(n==0)return 0;
        else if(n==1)return 1;
        unordered_map<int,int>map;
        while(i<n && j<n){
            if(map.find(s[i])!=map.end() and map[s[i]]>=j){
                j=map[s[i]]+1;
            }
            map[s[i]]=i;
            i++;
            ans=max(ans,i-j+1);
            
        }
        return --ans;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        cs=set()
        l=0
        res=0
        for r in range(len(s)):
            while s[r] in cs:
                cs.remove(s[l])
                l+=1
            cs.add(s[r])
            res=max(res,r-l+1)
        return res
        